lib/qiita/markdown/base_processor.rb

```diff
@@ -0,0 +1,43 @@
+module Qiita
+  module Markdown
+    # A base processor for rendering a markdown document.
+    # This class must be inherited and defined.
+    class BaseProcessor
+      # @param [Hash] context Optional context for HTML::Pipeline.
+      def initialize(context = {})
+        @context = self.class.default_context.merge(context)
+      end
+
+      # Converts Markdown text into HTML string with extracted metadata.
+      #
+      # @param [String] input Markdown text.
+      # @param [Hash] context Optional context merged into default context.
+      # @return [Hash] Process result.
+      # @example
+      #   Qiita::Markdown::Processor.new.call(markdown) #=> {
+      #     codes: [...],
+      #     mentioned_usernames: [...],
+      #     output: "...",
+      #   }
+      def call(input, context = {})
+        HTML::Pipeline.new(filters, @context).call(input, context)
+      end
+
+      # @note Modify filters if you want.
+      # @return [Array<HTML::Pipeline::Filter>]
+      def filters
+        @filters ||= self.class.default_filters.clone
```

Comment:
Because `.default_filters` always returns a new Array instance for each invocation, this `clone` is no longer needed.